### PR TITLE
Feature/concurrent dead lock

### PR DESCRIPTION
**Pull request by 244372610**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [x] Code compiles correctly.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by gaius-qi — APPROVED** [review]
LGTM

File: client/daemon/daemon.go [hunk 1]
```diff
@@ -111,9 +111,9 @@ func New(opt *config.DaemonOption) (Daemon, error) {
 				PeerId: request.PeerID,
 			})
 			if er != nil {
-				logger.Errorf("leave task %s/%s, error: %v", request.TaskID, request.PeerID, er)
+				logger.Errorf("step4:leave task %s/%s, error: %v", request.TaskID, request.PeerID, er)
```
Comment (jim3ma) [inline client/daemon/daemon.go:114]: `step4:leave ...` -> `step 4: leave ...`

File: client/daemon/daemon.go [hunk 2]
```diff
@@ -111,9 +111,9 @@ func New(opt *config.DaemonOption) (Daemon, error) {
...
 			} else {
-				logger.Infof("leave task %s/%s state ok", request.TaskID, request.PeerID)
+				logger.Infof("step4:leave task %s/%s state ok", request.TaskID, request.PeerID)
```
Comment (jim3ma) [inline client/daemon/daemon.go:116]: `step4:leave ...` -> `step 4: leave ...`

File: client/daemon/peer/peertask_file.go [hunk 3]
```diff
@@ -89,11 +89,13 @@ func newFilePeerTask(ctx context.Context,
 	// trace register
 	_, regSpan := tracer.Start(ctx, config.SpanRegisterTask)
 	result, err := schedulerClient.RegisterPeerTask(ctx, request)
+	logger.Infof("step1: peer %s start to register", request.PeerId)
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file.go:92]: `step1:` -> `step 1:`

File: client/daemon/peer/peertask_file.go [hunk 4]
```diff
@@ -89,11 +89,13 @@ func newFilePeerTask(ctx context.Context,
...
 	regSpan.RecordError(err)
 	regSpan.End()
 
 	var backSource bool
 	if err != nil {
+		logger.Errorf("step1: peer %s register failed: err", request.PeerId, err)
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file.go:98]: `step1:` -> `step 1:`

File: client/daemon/peer/peertask_file.go [hunk 5]
```diff
@@ -102,18 +104,17 @@ func newFilePeerTask(ctx context.Context,
...
 			span.End()
-			logger.Errorf("register peer task failed: %s, peer id: %s", err, request.PeerId)
 			return ctx, nil, nil, err
 		}
 	}
 	if result == nil {
 		defer span.End()
 		span.RecordError(err)
-		err = errors.Errorf("empty schedule result")
+		err = errors.Errorf("step1: peer register result is nil")
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file.go:113]: `step1:` -> `step 1:`

File: client/daemon/peer/peertask_file.go [hunk 6]
```diff
@@ -102,18 +104,17 @@ func newFilePeerTask(ctx context.Context,
...
 		return ctx, nil, nil, err
 	}
 	span.SetAttributes(config.AttributeTaskID.String(result.TaskId))
-	logger.Infof("register task success, task id: %s, peer id: %s, SizeScope: %s",
+	logger.Infof("step1: register task success, task id: %s, peer id: %s, SizeScope: %s",
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file.go:117]: `step1:` -> `step 1:`

File: client/daemon/peer/peertask_file.go [hunk 7]
```diff
@@ -147,7 +148,9 @@ func newFilePeerTask(ctx context.Context,
 	}
 
 	peerPacketStream, err := schedulerClient.ReportPieceResult(ctx, result.TaskId, request)
+	logger.Infof("step2: start report peer %s piece result", request.PeerId)
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file.go:151]: `step2:` -> `step 2:`

File: client/daemon/peer/peertask_file.go [hunk 8]
```diff
@@ -147,7 +148,9 @@ func newFilePeerTask(ctx context.Context,
...
 	if err != nil {
+		logger.Errorf("step2: peer %s report piece failed: err", request.PeerId, err)
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file.go:153]: `step2:` -> `step 2:`

File: client/daemon/peer/peertask_file_callback.go [hunk 9]
```diff
@@ -106,9 +106,9 @@ func (p *filePeerTaskCallback) Done(pt Task) error {
 		Code:           dfcodes.Success,
 	})
 	if err != nil {
-		pt.Log().Errorf("report successful peer result, error: %v", err)
+		pt.Log().Errorf("step3: report successful peer result, error: %v", err)
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file_callback.go:109]: `step3:` -> `step 3:`

File: client/daemon/peer/peertask_file_callback.go [hunk 10]
```diff
@@ -106,9 +106,9 @@ func (p *filePeerTaskCallback) Done(pt Task) error {
...
 	} else {
-		pt.Log().Infof("report successful peer result ok")
+		pt.Log().Infof("step3: report successful peer result ok")
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file_callback.go:111]: `step3:` -> `step 3:`

File: client/daemon/peer/peertask_file_callback.go [hunk 11]
```diff
@@ -131,9 +131,9 @@ func (p *filePeerTaskCallback) Fail(pt Task, code base.Code, reason string) erro
 		Code:           code,
 	})
 	if err != nil {
-		pt.Log().Errorf("report fail peer result, error: %v", err)
+		pt.Log().Errorf("step3: report fail peer result, error: %v", err)
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file_callback.go:134]: `step3:` -> `step 3:`

File: client/daemon/peer/peertask_file_callback.go [hunk 12]
```diff
@@ -131,9 +131,9 @@ func (p *filePeerTaskCallback) Fail(pt Task, code base.Code, reason string) erro
...
 	} else {
-		pt.Log().Infof("report fail peer result ok")
+		pt.Log().Infof("step3: report fail peer result ok")
```
Comment (jim3ma) [inline client/daemon/peer/peertask_file_callback.go:136]: `step3:` -> `step 3:`

File: client/daemon/peer/peertask_stream.go [hunk 13]
```diff
@@ -72,6 +72,7 @@ func newStreamPeerTask(ctx context.Context,
 	// trace register
 	_, regSpan := tracer.Start(ctx, config.SpanRegisterTask)
 	result, err := schedulerClient.RegisterPeerTask(ctx, request)
+	logger.Infof("step1: peer %s start to register", request.PeerId)
```
Comment (jim3ma) [inline client/daemon/peer/peertask_stream.go:75]: `step1:` -> `step 1:`

File: client/daemon/peer/peertask_stream.go [hunk 14]
```diff
@@ -128,6 +129,7 @@ func newStreamPeerTask(ctx context.Context,
 	}
 
 	peerPacketStream, err := schedulerClient.ReportPieceResult(ctx, result.TaskId, request)
+	logger.Infof("step2: start report peer %s piece result", request.PeerId)
```
Comment (jim3ma) [inline client/daemon/peer/peertask_stream.go:132]: `step2:` -> `step 2:`

**Review by jim3ma — APPROVED** [review]

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**